cmd/src/search_stream.go

```diff
@@ -0,0 +1,413 @@
+		apiFlags = api.StreamingFlags(flagSet)
+	)
+	if err := flagSet.Parse(args); err != nil {
+		return err
+	}
+
+	query := flagSet.Arg(0)
+
+	t, err := parseTemplate(streamingTemplate)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create request.
+	client := cfg.apiClient(apiFlags, flagSet.Output())
+	req, err := client.NewHTTPRequest(context.Background(), "GET", "search/stream?q="+url.QueryEscape(query), nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Accept", "text/event-stream")
+	if *display >= 0 {
+		q := req.URL.Query()
+		q.Add("display", strconv.Itoa(*display))
+		req.URL.RawQuery = q.Encode()
+	}
+
+	// Send request.
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error sending request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Process response.
+	err = streaming.Decoder{
+		OnProgress: func(progress *streaming.Progress) {
+			// We only show the final progress.
+			if !progress.Done {
+				return
+			}
+			err = t.ExecuteTemplate(os.Stdout, "progress", progress)
+			if err != nil {
+				fmt.Printf("ExecuteTemplate: %s\n", err)
+			}
+			return
+		},
+		OnError: func(eventError *streaming.EventError) {
+			fmt.Printf("ERR: %s", eventError.Message)
+		},
+		OnAlert: func(alert *streaming.EventAlert) {
+			proposedQueries := make([]ProposedQuery, len(alert.ProposedQueries))
+			for _, pq := range alert.ProposedQueries {
+				proposedQueries = append(proposedQueries, ProposedQuery{
+					Description: pq.Description,
+					Query:       pq.Query,
+				})
+			}
+
+			err = t.ExecuteTemplate(os.Stdout, "alert", searchResultsAlert{
+				Title:           alert.Title,
+				Description:     alert.Description,
+				ProposedQueries: proposedQueries,
+			})
+			if err != nil {
+				fmt.Printf("ExecuteTemplate: %s\n", err)
+				return
+			}
+		},
+		OnMatches: func(matches []streaming.EventMatch) {
+			for _, match := range matches {
+				if file, ok := match.(*streaming.EventFileMatch); ok {
+					err = t.ExecuteTemplate(os.Stdout, "file", struct {
+						Query string
+						*streaming.EventFileMatch
+					}{
+						Query:          query,
+						EventFileMatch: file,
+					},
+					)
+					if err != nil {
+						fmt.Printf("ExecuteTemplate: %s\n", err)
+						return
+					}
+					continue
+				}
+
+				if repo, ok := match.(*streaming.EventRepoMatch); ok {
+					err = t.ExecuteTemplate(os.Stdout, "repo", struct {
+						SourcegraphEndpoint string
+						*streaming.EventRepoMatch
+					}{
+						SourcegraphEndpoint: cfg.Endpoint,
+						EventRepoMatch:      repo,
+					})
+					if err != nil {
+						fmt.Printf("ExecuteTemplate: %s\n", err)
+						return
+					}
+					continue
+				}
+
+				if commit, ok := match.(*streaming.EventCommitMatch); ok {
+					err = t.ExecuteTemplate(os.Stdout, "commit", struct {
+						SourcegraphEndpoint string
+						*streaming.EventCommitMatch
+					}{
+						SourcegraphEndpoint: cfg.Endpoint,
+						EventCommitMatch:    commit,
+					})
```

Comment:
I think you could clean up this render code a bit if you introduce something like a `Context` struct that has all of the fields defined but on which you only set the current struct.

Reply:
Yeah, I also went back and forth about this one. I think the current approach is pretty straightforward although maybe a bit verbose. If we use a common context struct then the templates get more difficult to read. I will see if I can find something better.